RecoEgamma/EgammaTools/src/EgammaPCAHelper.cc

```diff
@@ -0,0 +1,342 @@
+        double mip = dEdXWeights_[layer] * 0.001;  // convert in GeV
+        if (thickness > 99. && thickness < 101)
+            mip *= invThicknessCorrection_[0];
+        else if (thickness > 199 && thickness < 201)
+            mip *= invThicknessCorrection_[1];
+        else if (thickness > 299 && thickness < 301)
+            mip *= invThicknessCorrection_[2];
+
+        pcavars[0] = recHitTools_->getPosition(rh_detid).x();
+        pcavars[1] = recHitTools_->getPosition(rh_detid).y();
+        pcavars[2] = recHitTools_->getPosition(rh_detid).z();
+        if (pcavars[2] == 0.)
+            std::cout << " Problem, hit with z =0 ";
+        else  {
+            Spot mySpot(rh_detid,itcheck->second->energy(),pcavars,layer,fraction,mip);
+            theSpots_.push_back(mySpot);
+        }
+    }
+    if (debug_) {
+        std::cout << " Stored " << theSpots_.size() << " hits " << std::endl;
+    }
+}
+
+void EGammaPCAHelper::computePCA(float radius , bool withHalo) {
+    // very important - to reset
+    pca_.reset(new TPrincipal(3, "D"));
+    bool initialCalculation = radius < 0;
+    if (debug_)
+        std::cout << " Initial calculation " << initialCalculation << std::endl;
+    if (initialCalculation && withHalo) {
+        std::cout << "Warning - in the first iteration, the halo hits are excluded " << std::endl;
+        withHalo=false;
+    }
+
+    float radius2 = radius*radius;
+    if (! initialCalculation)     {
+        math::XYZVector mainAxis(axis_);
+        mainAxis.unit();
+        math::XYZVector phiAxis(barycenter_.x(), barycenter_.y(), 0);
+        math::XYZVector udir(mainAxis.Cross(phiAxis));
+        udir = udir.unit();
+        trans_ = Transform3D(Point(barycenter_), Point(barycenter_ + axis_), Point(barycenter_ + udir), Point(0, 0, 0),
+        Point(0., 0., 1.), Point(1., 0., 0.));
+    }
+
+    unsigned nSpots = theSpots_.size();
+    std::set<int> layers;
+    for ( unsigned i =0; i< nSpots ; ++i) {
+        Spot spot(theSpots_[i]);
+        if (spot.layer() > 28) continue;
+        if (!withHalo && (! spot.isCore() ))
+            continue;
+        if (initialCalculation) {
+            // initial calculation, take only core hits
+            if ( ! spot.isCore() ) continue;
+            //std::cout << " Multiplicity " << spot.multiplicity() << " " << spot.row()[0] << " " ;
+            //std::cout << spot.row()[1] << " " << spot.row()[2] << std::endl;
+            layers.insert(spot.layer());
+            for (int i = 0; i < spot.multiplicity(); ++i)
+                pca_->AddRow(spot.row());
+        }
+        else { // use a cylinder, include all hits
+            math::XYZPoint local = trans_(Point( spot.row()[0],spot.row()[1],spot.row()[2]));
+            if (local.Perp2() > radius2) continue;
+            layers.insert(spot.layer());
+            for (int i = 0; i < spot.multiplicity(); ++i)
+                pca_->AddRow(spot.row());
```

Comment:
from the implementation it looks like a loop of size ```(int)energy/mip``` which repeatedly calls AddRow.
What is the typical size of spot.multiplicity?
If it's typically above 1, some different solution should be found.

Reply:
This is the big spot in the igprof right now, but we would have to self-implement a weighted PCA to avoid it.  This can be done, but perhaps for another PR?  We are already in the shadow of many other modules computing-wise.